src/orion/algo/evolution_es.py

```diff
@@ -0,0 +1,263 @@
+
+
+class BracketEVES(Bracket):
+    """Bracket of rungs for the algorithm Hyperband.
+
+    Parameters
+    ----------
+    evolutiones: `evolutiones` algorithm
+        The evolutiones algorithm object which this bracket will be part of.
+    budgets: list of tuple
+        Each tuple gives the (n_trials, resource_budget) for the respective rung.
+    repetition_id: int
+        The id of hyperband execution this bracket belongs to
+
+    """
+
+    def __init__(self, evolution_es, budgets, repetition_id, space):
+        super(BracketEVES, self).__init__(evolution_es, budgets, repetition_id)
+        self.eves = self.hyperband
+        self.space = space
+        self.search_space_remove_fidelity = []
+
+        if evolution_es.mutate_attr:
+            self.mutate_attr = copy.deepcopy(evolution_es.mutate_attr)
+        else:
+            self.mutate_attr = {}
+
+        function_string = self.mutate_attr.pop('function',
+                                               "orion.algo.mutate_functions.default_mutate")
+        mod_name, func_name = function_string.rsplit('.', 1)
+        mod = importlib.import_module(mod_name)
+        self.mutate_func = getattr(mod, func_name)
+
+        for i in range(len(space.values())):
+            if not i == self.eves.fidelity_index:
+                self.search_space_remove_fidelity.append(i)
+
+    def get_candidates(self, rung_id):
+        """Get a candidate for promotion"""
+        if self.has_rung_filled(rung_id + 1):
+            return []
+
+        rung = self.rungs[rung_id]['results']
+
+        population_range = (self.eves.nums_population
+                            if len(list(rung.values())) > self.eves.nums_population
+                            else len(list(rung.values())))
+
+        for i in range(population_range):
+            for j in self.search_space_remove_fidelity:
+                self.eves.population[j][i] = list(rung.values())[i][1][j]
+            self.eves.performance[i] = list(rung.values())[i][0]
+
+        population_index = list(range(self.eves.nums_population))
+        red_team = np.random.choice(population_index, self.eves.nums_comp_pairs, replace=False)
+        diff_list = list(set(population_index).difference(set(red_team)))
+        blue_team = np.random.choice(diff_list, self.eves.nums_comp_pairs, replace=False)
+
+        winner_list = []
+        loser_list = []
+
+        hurdles = 0
+        for i, _ in enumerate(red_team):
+            winner = (red_team
+                      if self.eves.performance[red_team[i]] < self.eves.performance[blue_team[i]]
+                      else blue_team)
+            loser = (red_team
+                     if self.eves.performance[red_team[i]] >= self.eves.performance[blue_team[i]]
+                     else blue_team)
+
+            winner_list.append(winner[i])
+            loser_list.append(loser[i])
+            hurdles += self.eves.performance[winner[i]]
+            self._mutate(winner[i], loser[i])
```

Comment:
It seems to me the second loop could be avoided by verifying here if the mutation created a duplicate in the population and then mutating again and again until it is not creating a duplicate.

Also, you could use `self.eves.get_id` (see [here]( https://github.com/Epistimio/orion/blob/develop/src/orion/algo/hyperband.py#L254)) to build a set of ids to test for duplicates.

Reply:
The second loop can be moved to 207, but can not be avoided. Because the we need to extract self.eves.population and assign it to points. The extraction process need one loop. get_id() need point object as its parameter rather than population, so we still need to extract population to point firstly. I tried to do that following by your suggestion, but it can not save much space or branches. I didn't make change here, because the flow: rungs-->population-->points is clear.